assembly/src/parsers/mod.rs

```diff
@@ -51,6 +51,10 @@ pub struct ProgramAst {
 }
 
 impl ProgramAst {
+    pub fn new(local_procs: Vec<ProcedureAst>, body: Vec<Node>) -> Self {
+        Self { local_procs, body }
+    }
```

Comment:
A couple of comments:

1. Since this is a part of a public interface, let's add doc comments to this function.
2. I'd probably move error checking logic which is currently on lines 119-121 to here and return `Result<ProgramAst, ParsingError>` rather than just `ProgramAst`.

Reply:
Done, though there is an issue with the deserialization functions because they return `Result<Self, DeserializationError>` rather than `Result<Self, ParsingError>`. I've added a conversion from the former to the latter type, but I'm not sure this is what we want.

In particular, the santity check that was moved from the parser to the constructor is not performed by the deserializer, which suggests that we can trust the serialized format?

Reply:
I think the types of check were are now doing were implicit in how the deserializer worked (i.e., the number of procedures could not exceed `u16::MAX` because to determine the number of procedures the deserializer read a `u16` value from the byte stream.

So, technically, these checks are somewhat redundant for the deserializer - but I don't think they hurt.